pkg/chunkenc/symbols.go

```diff
@@ -0,0 +1,374 @@
+package chunkenc
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"hash"
+	"io"
+	"sync"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/model/labels"
+
+	"github.com/grafana/loki/pkg/util"
+)
+
+// symbol holds reference to a label name and value pair
+type symbol struct {
+	Name, Value uint32
+}
+
+type symbols []symbol
+
+// symbolizer holds a collection of label names and values and assign symbols to them.
+// symbols are actually index numbers assigned based on when the entry is seen for the first time.
+type symbolizer struct {
+	mtx            sync.RWMutex
+	symbolsMap     map[string]uint32
+	labels         []string
+	size           int
+	compressedSize int
+}
+
+func newSymbolizer() *symbolizer {
+	return &symbolizer{
+		symbolsMap: map[string]uint32{},
+	}
+}
+
+// Add adds new labels pairs to the collection and returns back a symbol for each existing and new label pair
+func (s *symbolizer) Add(lbls labels.Labels) symbols {
+	if len(lbls) == 0 {
+		return nil
+	}
+
+	syms := make([]symbol, len(lbls))
+
+	for i, label := range lbls {
+		syms[i].Name = s.add(label.Name)
+		syms[i].Value = s.add(label.Value)
+	}
+
+	return syms
+}
+
+func (s *symbolizer) add(lbl string) uint32 {
+	s.mtx.RLock()
+	idx, ok := s.symbolsMap[lbl]
+	s.mtx.RUnlock()
+
+	if ok {
+		return idx
+	}
+
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	idx, ok = s.symbolsMap[lbl]
+	if !ok {
+		idx = uint32(len(s.symbolsMap))
```

Comment:
Since `idx` is the index within the `labels` slice, I think it would be easier to understand this if we look at the len of `labels`

Reply:
It should have the same length. `symbolsMap` is actually for deduping while `labels` is used at query time to find labels referenced by index. I will change it the way you suggested since I am fine with either ways.